Generic Primitive Drawing

RobertBColton opened the pull request:
This cleanup has been coming for a while, but I am glad to finally do it. This moves all of the `draw_primitive_*` and `d3d_primitive_*` functions to `Graphics_Systems/General/GSprimitives.cpp` where we've always wanted them. They are implemented on top of the generic `d3d_model_*` functions now. This means the following:

* GL1 doesn't use actual immediate mode, it will use a VBO or a vertex array depending on what fallback the `vertex_*` functions are using. Regardless, it will still be GL1.1 compliant. Hopefully much faster after benchmarking because immediate mode is slow compared even to vertex arrays.
* GL3 and DX9 don't need to do the batching themselves now, and the 3 headers in `Bridges/General` can just be deleted later.
* Where ENIGMA doesn't implicitly flush the primitive batch, the user can do so by calling `draw_batch_flush` to draw all current primitives and then clear the primitives model.

### Optimization
Obviously I am going to do some benchmarking here to ensure this doesn't degrade performance and tune the pull request before it is reviewed. This should have the same or better performance everywhere.

Master
======

##### `draw_sprite` benchmark
| Branch     | FPS | CPU RAM (MB) | GPU RAM (MB) |
|------------|-----|--------------|--------------|
| GMS v1.4   | 645 | 28.8         | 22           |
| ENIGMA GL1 | 242 | 44.0         | 13           |
| ENIGMA GL3 | 373 | 56.5         | 34           |
| ENIGMA DX9 | 384 | 17.3         | 26           |

##### `draw_rectangle` outlined benchmark
| Branch     | FPS | CPU RAM (MB) | GPU RAM (MB) |
|------------|-----|--------------|--------------|
| GMS v1.4   | 25  | 34.9         | 22           |
| ENIGMA GL1 | 79  | 27.2         | 13           |
| ENIGMA GL3 | 17  | 74.5         | 29           |
| ENIGMA DX9 | 68  | 17.4         | 22           |

Pull request unoptimized
======
##### `draw_sprite` benchmark
| Branch                   | FPS | CPU RAM (MB)  | GPU RAM (MB) |
|--------------------------|-----|---------------|--------------|
| ENIGMA GL1 Vertex Arrays | :small_red_triangle_down: 118 | :small_red_triangle_down: 45.5          | :heavy_minus_sign: 13           |
| ENIGMA GL1 VBO           | :small_red_triangle_down: 026 | :small_red_triangle_down: 55.0 (-/+ 10) | :small_red_triangle_down: 36           |
| ENIGMA GL3               | :small_red_triangle_down: 026 | :small_red_triangle_down: 70.0 (-/+ 10) | :heavy_check_mark: 28           |
| ENIGMA DX9               | :small_red_triangle_down: 042 | :heavy_check_mark: 14.6          | :heavy_minus_sign: 26           |

##### `draw_rectangle` outlined benchmark
| Branch                   | FPS | CPU RAM (MB) | GPU RAM (MB) |
|--------------------------|-----|--------------|--------------|
| ENIGMA GL1 Vertex Arrays | :small_red_triangle_down: 66  | :small_red_triangle_down: 27.6         | :small_red_triangle_down: 28           |
| ENIGMA GL1 VBO           | :small_red_triangle_down: 15  | :small_red_triangle_down: 61.3 (-/+ 2) | :small_red_triangle_down: 22           |
| ENIGMA GL3               | :small_red_triangle_down: 15  | :heavy_check_mark: 65.0 (-/+ 8) | :small_red_triangle_down: 36           |
| ENIGMA DX9               | :small_red_triangle_down: 34  | :heavy_check_mark: 14.9         | :small_red_triangle_down: 23           |

Deferred flushing
======
This is an obvious optimization that GMS makes as well. We can defer any flushing of the batch until a state change occurs or just before the scene ends. This is a bit tricky to get right, but I believe what I have is good. Below I have a list of all of the ways a batch flush can be triggered implicitly in this mode.

##### `draw_sprite` benchmark
| Branch                   | FPS | CPU RAM (MB) | GPU RAM (MB) |
|--------------------------|-----|--------------|--------------|
| ENIGMA GL1 Vertex Arrays | :heavy_check_mark: 313 | :heavy_check_mark: 40.9         | :small_red_triangle_down: 69           |
| ENIGMA GL1 VBO           | :heavy_check_mark: 334 | :heavy_check_mark: 39.4         | :small_red_triangle_down: 40           |
| ENIGMA GL3               | :small_red_triangle_down: 336 | :heavy_check_mark: 51.2         | :heavy_check_mark: 11           |
| ENIGMA DX9               | :small_red_triangle_down: 349 | :heavy_minus_sign: 17.3         | :heavy_check_mark: 23           |

##### `draw_rectangle` outlined benchmark
| Branch                   | FPS | CPU RAM (MB) | GPU RAM (MB) |
|--------------------------|-----|--------------|--------------|
| ENIGMA GL1 Vertex Arrays | :small_red_triangle_down: 77  | :heavy_check_mark: 25.0         | :small_red_triangle_down: 16           |
| ENIGMA GL1 VBO           | :small_red_triangle_down: 73  | :small_red_triangle_down: 39.2         | :small_red_triangle_down: 22           |
| ENIGMA GL3               | :heavy_check_mark: 79  | :heavy_check_mark: 50.7         | :small_red_triangle_down: 35           |
| ENIGMA DX9               | :heavy_check_mark: 76  | :heavy_check_mark: 17.3         | :small_red_triangle_down: 24           |

Implicit flushing triggers
======
This is a list of all the ways in which a batch flush can be triggered implicitly by ENIGMA and in what mode it occurs. In all of the cases, the batch will only be flushed if it is actually dirty. We also guard against flushing while an active flush is going on.

* When the user changes the batch mode with `draw_set_batch_mode` (always)
* At the end of all draw events (deferred)
* At the end of all draw GUI events, if GUI events are used (deferred)
* Just after drawing `cursor_sprite` and the scene ends (deferred)
* Just before the scene actually ends (never mode/prevent leaks)
* When the user calls `vertex_submit` or `index_submit` to draw their own vertex buffer (deferred)
* When a new 2D or 3D primitive is started and the texture is different or the user wants to draw without a texture (deferred)
* When a 2D or 3D primitive ends (immediate)
* When one of the `d3d_draw_*` shape functions is called and the texture is different (deferred)
* Just before one of the `d3d_draw_*` shape functions returns (immediate)
* Before any rendering state (e.g, draw_set_blend_mode, d3d_set_culling) is changed (deferred)
* Any texture state changes or changing of the target texture (deferred)
* Any changing or copying of a render target or surface (deferred)
* When any of the `d3d_set_projection_*` or `d3d_transform_*` functions is called (deferred)
* Just before clearing of the color, depth, or stencil buffers (deferred)
* Just before saving a screenshot or changing the viewport (deferred)
* Any change in uniform state or the currently bound shader program (deferred)

RobertBColton commented:
This time, EnigmaBot is correct that the shape functions have changed. Here is GM8 and GMS v1.4 for comparison.

GM8 | GMS v1.4
--- | ---
![GM8 3D Shapes Test](https://user-images.githubusercontent.com/3212801/44296187-4bc0a080-a287-11e8-8149-7d091e2295c4.png) | ![GMS v1.4 3D Shapes Test](https://user-images.githubusercontent.com/3212801/44296197-6bf05f80-a287-11e8-9faf-0c135c8e64bd.png)

RobertBColton commented on ENIGMAsystem/SHELL/Graphics_Systems/General/GSprimitives.cpp:
```diff
@@ -1,17 +1,234 @@
+  if (!enigma_user::d3d_model_exists(draw_batch_stream))
+    draw_batch_stream = enigma_user::d3d_model_create(enigma_user::model_stream);
+  return draw_batch_stream;
+}
```
@JoshDreamland when you review this for me, please look over this particular spot very carefully because I was running into undefined behavior before I changed it to this (61ed4be743191ec2fa2473170c934f0b160da9b1).

JoshDreamland replied:
I did not recognize that this was static. Statics are initialized the first time the function is run. It should not behave differently now unless someone was destroying the resource this method returns. Is that the case? Also, is this method being called before `main()`?

RobertBColton replied:
As far as I know, no, nobody was destroying it before. I don't believe it is being called before main. It would be called anywhere `draw_primitive_begin/_end` is called or `d3d_draw_*`. So it should only be getting called inside of a draw event.

JoshDreamland replied:
If you print every time you allocate a new buffer (every time you enter that `if`),  how often does it print?

RobertBColton replied:
Apparently, just a single time.

```cpp
  if (!enigma_user::d3d_model_exists(draw_batch_stream)) {
    draw_batch_stream = enigma_user::d3d_model_create(enigma_user::model_stream);
    std::cerr << "new batch stream model created: " << darw_batch_stream << std::endl;
  }
```
![MSYS2 Log for Batch Stream Model Create](https://user-images.githubusercontent.com/3212801/44305402-86d4d980-a344-11e8-943e-c0c9cbc19935.png)

RobertBColton replied:
```cpp
int draw_get_batch_stream() {
  static int draw_batch_stream = enigma_user::d3d_model_create(enigma_user::model_stream);
  return draw_batch_stream;
}
```
Yes, it goes back to the weird FPS dropping when focused issue again. Strange indeed.

JoshDreamland replied:
Print the IDs in both cases, please.

RobertBColton replied:
@JoshDreamland alright, this is mildly depressing. If, after undoing the above shit, I enable link-time optimization for run mode? It results in the same FPS drop to 150fps in Direct3D9:
```cpp
	override CXXFLAGS += -Wall -s -O3 -fno-rtti -fno-exceptions -flto
	override CFLAGS += -Wall -s -O3 -flto -fno-exceptions
	override LDFLAGS += -flto -s -O3
```

We're getting bit somewhere, and I don't like it.

RobertBColton replied:
I can reproduce that on master too all the way back 2016, I spent a whole day trying to find it. I am going to just put it aside for now and hope that you'll do me a solid some weekend and run Valgrind against master.

JoshDreamland commented on ENIGMAsystem/SHELL/Graphics_Systems/General/GSprimitives.cpp:
```diff
@@ -1,17 +1,234 @@
+    case pr_trianglestrip: if (vertex_count > 2) return vertex_count - 2; break;
+    case pr_trianglefan: if (vertex_count > 2) return vertex_count - 2; break;
+  }
+  return 0;
```
Add a debug-mode assertion fail, here (it's good to know you passed an invalid argument).

RobertBColton replied:
Alright: e661c9e5f3c37f4edeb6cf17fe22dbf7e174f11b

I agree that specifying an unhandled primitive kind is an error. I did not, however, add a message for when you specify a vertex count that is too low because that's not really an error. You may have a single vertex etc and still want to know if that's enough to draw n primitives, which is the point of this function.

RobertBColton commented:
Alright, good, the bot is back to not being full of shit. That's correct, I've fixed it to pass that draw test. The transform functions should have been doing a deferred flush, which they are now.

RemoveRusky commented:
pls quit huding like bitch @RobertBColton

JoshDreamland commented on ENIGMAsystem/SHELL/Graphics_Systems/Direct3D11/DX11colors.cpp:
```diff
@@ -85,6 +88,7 @@ void draw_set_color_rgba(unsigned char red,unsigned char green,unsigned char blu
 void draw_set_color_write_enable(bool red, bool green, bool blue, bool alpha)
 {
-
+	draw_batch_flush(batch_flush_deferred);
```
Doesn't this function control whether we're writing colors into a VBO? Why does this require a flush?

RobertBColton replied:
This controls the color buffer (depth buffer's best friend) of the framebuffer (nothing to do with the vertex data). If you turn off writing to e.g, the red channel of the color buffer, then you need to flush everything that was already drawn just before making the state change.

Please review the documentation for glColorMask:
https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glColorMask.xhtml

JoshDreamland commented on ENIGMAsystem/SHELL/Graphics_Systems/Direct3D11/DX11vertex.cpp:
```diff
@@ -248,6 +249,8 @@ void vertex_color(int buffer, int color, double alpha) {
 }
 
 void vertex_submit(int buffer, int primitive, unsigned start, unsigned count) {
+  draw_batch_flush(batch_flush_deferred);
```
Please reevaluate whether this function requires a flush (I am not certain myself).

RobertBColton replied:
It is needed, consider the following scenario:
* The user sets the batch flush mode to deferred (won't flush until state change)
* The user calls `draw_sprite` a few times (which is batched together but waiting to be drawn)
* The user then calls `vertex_submit` without a texture (no state change occurs) and the vertex buffer is drawn immediately
* The `draw_sprite` calls still haven't been flushed... and if they are flushed beyond this point, they will appear "above" the vertex buffer the user just drew

JoshDreamland commented on ENIGMAsystem/SHELL/Graphics_Systems/Direct3D9/DX9vertex.cpp:
```diff
@@ -197,11 +197,9 @@ void vertex_color(int buffer, int color, double alpha) {
 
 void vertex_submit(int buffer, int primitive, unsigned start, unsigned count) {
-  const enigma::VertexBuffer* vertexBuffer = enigma::vertexBuffers[buffer];
+  draw_batch_flush(batch_flush_deferred);
```
(For all of these.)

RobertBColton replied:
They do, read my comment above.

JoshDreamland commented on ENIGMAsystem/SHELL/Graphics_Systems/General/GSprimitives.cpp:
```diff
@@ -1,17 +1,254 @@
+{
+  draw_batch_begin_deferred(texId);
+  d3d_model_block(draw_get_batch_stream(), x1, y1, z1, x2, y2, z2, hrep, vrep, closed);
+  draw_batch_flush(batch_flush_immediate);
```
Why are you flushing this each time? Doesn't that defeat the point of this batching? Or is that the point of this batch mode flag you are passing?

RobertBColton replied:
That is the point of the `batch_flush_immediate` flag, the flush will return early if we're not in immediate batch mode. So in deferred mode, no, it is not flushed each time.

JoshDreamland commented on ENIGMAsystem/SHELL/Graphics_Systems/General/GSscreen.cpp:
```diff
@@ -267,9 +270,19 @@ void screen_redraw()
+
+  if (sprite_exists(cursor_sprite)) {
     draw_sprite(cursor_sprite, 0, mouse_x, mouse_y);
+    // cursor sprite needs its own flush before this frame ends
```
Why? The frame ending will flush, will it not?

RobertBColton commented on ENIGMAsystem/SHELL/Graphics_Systems/Direct3D9/DX9draw.cpp:
```diff
@@ -50,10 +49,11 @@ bool draw_get_msaa_supported()
 void draw_set_msaa_enabled(bool enable)
 {
-
+	draw_batch_flush(batch_flush_deferred);
```
I stopped doing that about halfway through this pr because I already realized my next pr, the render state cache, is going to delete all of this anyway. Sorry.

RobertBColton commented on ENIGMAsystem/SHELL/Graphics_Systems/Direct3D9/DX9matrix.cpp:
```diff
@@ -15,16 +15,21 @@
+namespace {
+
+void graphics_set_transform(D3DTRANSFORMSTATETYPE State, const D3DMATRIX *pMatrix) {
+  draw_batch_flush(batch_flush_deferred);
```
I'm going to delete `d3dmgr->SetTransform` in my next pull request, hence I want to start removing calls to it now.

`DX9Context.h`, `GL3Context.h`, and `DX11Context.h` are going to be replaced by `Graphics_Systems/General/GSstate.cpp` and maybe `GSstate.h`.

Separately, I also plan to deduplicate this file by implementing `matrix_build` and `matrix_set` which will further deduplicate all of this code. At that time, then `graphics_set_transform` can simply become the `matrix_set` GMS function. You'll see, I have no further comments on this right now.

RobertBColton commented on ENIGMAsystem/SHELL/Graphics_Systems/Direct3D9/DX9matrix.cpp:
```diff
@@ -38,7 +43,7 @@ void d3d_set_perspective(bool enable)
   }
   //set projection matrix
-  d3dmgr->SetTransform(D3DTS_PROJECTION, &matProjection);
+  graphics_set_transform(D3DTS_PROJECTION, &matProjection);
```
`graphics_set_transform` will eventually become the user function `matrix_set` and be implemented generically. You'll see, it's for a future pr.

RobertBColton commented on ENIGMAsystem/SHELL/Graphics_Systems/OpenGL3/GL3shader.cpp:
```diff
@@ -642,7 +642,7 @@ void glsl_program_detach(int id, int sid)
 void glsl_program_set(int id)
 {
   if (enigma::bound_shader != id){
-    oglmgr->ShaderFunc();
```
Each of the `oglmgr` functions I removed was just for flushing the batch, that's all they did, and some of them didn't actually do that. They were just left empty. I don't remember if that was one of them. The reason I am so interested in doing away with these and creating `GSstate.h`/`GSstate.cpp` in a future pr is because I don't feel we should have a lot of graphics code in `Bridges/General/` because that makes no sense. I also feel that the logic should be deduplicated, see my other comments below about my plans. Finally, I feel that the user should be given more control over the whole thing (hence the niche & novel features in this pull request).

RobertBColton commented on ENIGMAsystem/SHELL/Graphics_Systems/General/GSprimitives.cpp:
```diff
@@ -1,17 +1,254 @@
+
+void draw_primitive_begin_texture(int kind, int texId)
+{
+  draw_batch_begin_deferred(texId);
```
@JoshDreamland The actual batching logic is here, when you begin a primitive. It decides to break the current batch if the texture is different, or else continue combining primitives into the batch model. It does that for deferred mode. In immediate mode it just flushes when the primitive ends.

RobertBColton commented on ENIGMAsystem/SHELL/Graphics_Systems/Direct3D9/DX9surface.cpp:
```diff
@@ -150,7 +146,6 @@ int surface_getpixel(int id, int x, int y)
   if (y < 0) y = 0;
   if (x > surface->width || y > surface->height) return 0;
 
-  d3dmgr->EndShapesBatching();
```
I was not, hence it's back: 154375f5b36412d2ba17afb55444a4832dd391f4

Reason I did that was because I was on the thought process that reading the surface, while still rendering to it, was undefined. Now that I've put more thought into it, not defining that is problematic and definitely less compatible.

Side note, I want to clean up all of the duplicate surface code at some point. All they are is a composite of different kinds of textures and they can very easily be rebuilt on top of an abstraction. Another issue is that we don't necessarily want to flush unless you are requesting data from the currently bound surface (would be more efficient than what we're doing now, but this is still essentially a no-op if the batch isn't dirty anyway). We'll have to come back to this later.

RobertBColton replied:
I should also mention, that I believe these surface changes are what made HitCoder's Sonic game start working in Direct3D9 on this pull request. So all around, I think we have definitely improved things here by making the batching more consistent across the systems.
![HitCoder's Sonic Game in Direct3D9](https://user-images.githubusercontent.com/3212801/44426231-13a5b000-a55c-11e8-9b88-345d88a2ad97.png)

RobertBColton commented on ENIGMAsystem/SHELL/Graphics_Systems/OpenGL3/GL3d3d.cpp:
```diff
@@ -73,7 +74,7 @@ namespace enigma_user
 {
 
 void d3d_clear_depth(double value) {
-  oglmgr->BlendFunc();
```
You may have a point here, but there's another side of this too that I am thinking about for future prs. Above I talked to you about how we may later want to only flush the batch for `surface_getpixel` if the id of the surface requested is currently bound. That is an excellent case of where I want more information. But I want to deduplicate the surface code and I want to implement a render state cache. None of the old functions I removed here are actually providing me the needed information anyway, so best just hold that thought for the future and not conflate issues here.

RobertBColton commented:
Yeah, yeah, yeah @EnigmaBot, we've heard enough of your shit.

RobertBColton commented:
Alright, I came up with a solution (00ff9812ed14cf6efea4064ee4e9f03a70109bb8) to the 3D shapes needing the current color and alpha problem that I think is actually pretty great. Instead of having `GSprimitives.cpp` deal with the current color, I have added a special parameter to `d3d_model_create` to make it automatically store the current color and alpha for each vertex. The batch streaming model is created with this parameter set to true.

![Wild Racing 3D Shapes Using Current Alpha](https://user-images.githubusercontent.com/3212801/44434473-4f02a780-a579-11e8-932a-3be40c030930.png)

A model that uses the draw color will add the current color and alpha just before adding of the second vertex or on model primitive end. The reason I chose to do it this way is because all of GM's vertex specification functions have always put color as the last argument. Hence, we should put the model in a format consistent with expectations.

You'll notice the documentation, even now in GMS, has also always mentioned the color last for functions that do not accept a color.
https://docs.yoyogames.com/source/dadiospice/002_reference/drawing/drawing%20primitives/draw_vertex_texture.html

RobertBColton commented:
Ok, so I caught one other final issue related to this but actually coming from the new models class. The issue was with mixing `draw_primitive` and `d3d_primitive` in the same batch. I wrote the new model class to allow primitives to have varying formats in the same model. I've already fixed it, all I had to do was provide a `vertex_submit` that takes a byte offset into the buffer instead of a starting index. For `vertex_submit`, the `start` argument is dependent on a consistent vertex format but the `offset` argument is not. I think this is a great solution because it gives the user many options and overloads to easily work with ranges of vertex buffers (which YoYo doesn't even allow).

I also decided to change the C namespace of some of the functions to avoid ambiguity in overload (because C/C++) and also to match GML API better. I also changed the equivalent index buffer functions to match. I did not provide an `index_submit_offset` yet because we don't quite have a need and I didn't want to conflate issues here.
```cpp
void vertex_submit(int buffer, int primitive);
void vertex_submit(int buffer, int primitive, int texture);
void vertex_submit_range(int buffer, int primitive, unsigned start, unsigned count);
void vertex_submit_range(int buffer, int primitive, int texture, unsigned start, unsigned count);
void vertex_submit_offset(int buffer, int primitive, unsigned offset, unsigned start, unsigned count);
void vertex_submit_offset(int buffer, int primitive, int texture, unsigned offset, unsigned start, unsigned count);

void index_submit(int buffer, int vertex, int primitive);
void index_submit(int buffer, int vertex, int primitive, int texture);
void index_submit_range(int buffer, int vertex, int primitive, unsigned start, unsigned count);
void index_submit_range(int buffer, int vertex, int primitive, int texture, unsigned start, unsigned count);
```

This essentially resolves the last of the issues with this or any of the other recent changes. As you can see from EnigmaBot's comment above, I've also updated the draw test to include a test for mixed vertex formats. @JoshDreamland this is ready to go!
